crates/accelerate/src/quantum_circuit/intern_context.rs

```diff
@@ -0,0 +1,144 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use std::collections::hash_map::DefaultHasher;
+use std::collections::VecDeque;
+use std::hash::{Hash, Hasher};
+use std::mem::take;
+
+macro_rules! println {
+    ($($rest:tt)*) => {
+        #[cfg(debug_interner)]
+        std::println!($($rest)*)
+    }
+}
+
+#[cfg(debug_interner)]
+fn unique_id() -> u64 {
+    use std::sync::atomic::AtomicU64;
+    use std::sync::atomic::Ordering::SeqCst;
+    static COUNTER: AtomicU64 = AtomicU64::new(0);
+    COUNTER.fetch_add(1, SeqCst)
+}
+
+fn hash<T>(vec: &Vec<T>) -> u64
+where
+    T: Hash,
+{
+    let mut hasher = DefaultHasher::default();
+    vec.hash(&mut hasher);
+    hasher.finish()
+}
+
+pub type IndexType = u16;
+pub type BitType = u32;
+
+#[derive(Clone, Debug)]
+struct SharedOperandList {
+    operands: Vec<BitType>,
+    use_count: usize,
+}
+
+#[pyclass(module = "qiskit._accelerate.quantum_circuit")]
+#[derive(Clone, Debug)]
+pub struct InternContext {
+    slots: Vec<Option<SharedOperandList>>,
+    free_slots: VecDeque<IndexType>,
+    slot_lookup: HashMap<u64, IndexType>,
+    #[cfg(debug_interner)]
+    id: u64,
+}
+
+impl InternContext {
+    pub fn intern(&mut self, args: Vec<BitType>) -> Option<IndexType> {
+        if self.free_slots.is_empty() && self.slots.len() == IndexType::MAX.into() {
+            return None;
+        }
+
+        let args_hash = hash(&args);
+        let slot_idx = self.slot_lookup.entry(args_hash).or_insert_with(|| {
```

Comment:
Why are we hashing a hash?  This doesn't feel right - two things could hash to the same value, and then there'd be an unresolvable collision.

Reply:
Ah, I think you're saying just use the `Vec<u32>` as the key for the lookup directly? I don't remember exactly why I didn't do that. It may have been because in an earlier revision I was using a 32 bit hasher to save space, but in the current implementation that's negligible.

Reply:
(also why DoS dictionary attacks work on unsalted hashes - if you deliberately construct data that all hashes to the same thing but isn't equal, the assumptions that make hash tables have amortised O(1) lookup fail into O(n))

Reply:
Makes sense, thanks! 😄

Reply:
Fixed in e1faea3. I ended up needing to use `Arc` as the owner of the `Vec<BitType>` to avoid keeping two copies between the `SharedOperandList` and the `slot_lookup` map.